### PR TITLE
feat(lsp): Implement textDocument/prepareCallHierarchy

### PR DESCRIPTION
**Pull request by jeanp413**
Also implements `callHierarchy/incomingCalls` and `callHierarchy/outgoingCalls`

Related #8643

### CODE REVIEW

**Comment by jeanp413** [comment]
Don't know why my test is failing in the CI, it runs fine locally, any ideas?

**Comment by lucacasonato** [comment]
cc @kt3k for the CI failure

File: cli/lsp/tsc.rs [hunk 1]
```diff
@@ -917,6 +951,182 @@ impl ReferenceEntry {
...
+
+    lsp::CallHierarchyItem {
+      name,
+      tags,
+      uri,
+      detail: Some(detail),
+      kind: self.kind.clone().into(),
+      range: self.span.to_range(line_index),
+      selection_range: self.selection_span.to_range(line_index),
+      data: None,
```
Comment (kitsonk) [inline cli/lsp/tsc.rs:1050]:
```suggestion
      ..Default::default()
```

Like I used in the other PR, this helps just ensure we have less "challenges" maintaining code when there are potential changes to the protocol in the future.

**Review by kitsonk — COMMENTED** [review]
One minor comment... thanks for all your contributions.

File: cli/lsp/tsc.rs (hunk 1 above)
Comment (jeanp413) [reply cli/lsp/tsc.rs:1050]: This won't work as `lsp::CallHierarchyItem` does not implement the `Default` trait

**Review by kitsonk — APPROVED** [review]
LGTM

**Pull request merged**